crates/sparrow-merge/src/merge_pipeline.rs

```diff
@@ -0,0 +1,340 @@
+            consumers,
+            left,
+            right,
+            result_type: result_type.clone(),
+        })
+    }
+
+    fn complete_partition(
+        &self,
+        input_partition: Partition,
+        merge_partition: &MergePartition,
+        scheduler: &mut dyn Scheduler,
+    ) -> error_stack::Result<(), PipelineError> {
+        tracing::info!(
+            "Input is closed and empty. Closing consumers and finishing partition {}.",
+            input_partition
+        );
+        self.consumers.close_input(input_partition, scheduler)?;
+        merge_partition.task.complete();
+        Ok(())
+    }
+}
+
+impl Pipeline for MergePipeline {
+    fn initialize(&mut self, tasks: Partitioned<TaskRef>) {
+        self.partitions = tasks
+            .into_iter()
+            .map(|task| {
+                let (tx_l, rx_l) = tokio::sync::mpsc::unbounded_channel();
+                let (tx_r, rx_r) = tokio::sync::mpsc::unbounded_channel();
+                MergePartition {
+                    task,
+                    is_closed: AtomicBool::new(false),
+                    txs: vec![tx_l, tx_r],
+                    // TODO: Interpolation
+                    // Current impl uses unlatched spread (`Null` interpolation).
+                    handler: Mutex::new(MergePartitionHandler {
+                        rxs: vec![rx_l, rx_r],
+                        merger: HeterogeneousMerge::new(
+                            &self.result_type,
+                            &self.left.datatype,
+                            &self.right.datatype,
+                        ),
+                    }),
+                }
+            })
+            .collect();
+    }
+
+    fn add_input(
+        &self,
+        input_partition: Partition,
+        input: usize,
+        batch: Batch,
+        scheduler: &mut dyn Scheduler,
+    ) -> error_stack::Result<(), PipelineError> {
+        error_stack::ensure!(
+            input == 0 || input == 1,
+            PipelineError::InvalidInput {
+                input,
+                input_len: 2
+            }
+        );
+
+        let partition = &self.partitions[input_partition];
+        error_stack::ensure!(
+            !partition.is_closed(),
+            PipelineError::InputClosed {
+                input,
+                input_partition
+            }
+        );
+
+        partition.txs[input]
+            .send(batch)
+            .into_report()
+            .change_context(PipelineError::Execution)?;
+        scheduler.schedule(partition.task.clone());
+        Ok(())
+    }
+
+    fn close_input(
+        &self,
+        input_partition: Partition,
+        input: usize,
+        scheduler: &mut dyn Scheduler,
+    ) -> error_stack::Result<(), PipelineError> {
+        error_stack::ensure!(
+            input == 0 || input == 1,
+            PipelineError::InvalidInput {
+                input,
+                input_len: 2
+            }
+        );
+
+        {
```

Comment:
Feels like even if you required this only has 2 elements, it would be cleaner if we used a vector here. Then you could do `*self.inputs[input].is_closed.lock() = true`. Maybe that's a good first step -- move away from `left/right` and these conditionals, towards a 2 element vector (and change assertions to use `input < self.inputs.len()`, etc. Then we'll be closer to making this support multiple inputs, without needing to do all of it right away.

Reply:
Probably so -- but I'm not sure yet where the "do n-way merge" logic will live. Will it be handled in `heterogenous merge`? Or just in the pipeline here, calling merge on specific batches each time? 

hetergeneous merge is written to handle `left` and `right` atm, so I may leave this to avoid complicating that code right now